injection/sharedmain/main.go

```diff
@@ -144,7 +153,22 @@ func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, cto
 
 	// Start all of the controllers.
 	logger.Info("Starting controllers...")
-	controller.StartAll(ctx.Done(), controllers...)
+	go controller.StartAll(ctx.Done(), controllers...)
+
+	profilingServer := profiling.NewServer(profilingHandler)
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(profilingServer.ListenAndServe)
+
+	// This will block until either a signal arrives or one of the grouped functions
+	// returns an error.
+	<-egCtx.Done()
+
+	profilingServer.Shutdown(context.Background())
```

Comment:
The way this is written, I'm not sure how the profiling server is supposed to shutdown when a signal comes in?

I'm also not sure what the singleton errorgroup is getting us?

Reply:
It's a bit... intricate. But it should work.

The errGroup Context is canceled once the signal comes in (it propagates through it) or once one of the eg functions finish. So in this case, we pass the `<-egCtx.Done()` line either if the profilingServer crashes or if the signal comes in.

Reply:
I think the errgroup brings us a little cleaner design. We don't need to define another channel for getting errors from the profilingServer and then wait for two channels like it is done in cmd/activator/main.go

@mattmoor regarding the server shutdown, I'm using a similar approach to activator and autoscaler where just `server.Shutdown(context.Background())` is called. If we want to catch signals then I think we'd need to pass `signals.NewContext()` instead of the Background context. In all the places. WDYT?

Reply:
Yeah, I'll bite my tongue on the go stdlib's poor context integration (IMO). ;-)

Reply:
I don't think we want to catch signals there though. A SIGKILL will kill the thing forcefully anyway and at SIGTERM we're kinda supposed to drain so this is imo fine as is.